osquery/tables/system/windows/windows_search.cpp

```diff
@@ -0,0 +1,478 @@
+    // TODO: find a way to verify that the columns requested exist before the
+    // query else we just get an error. If a new column is added or dropped on
+    // an OS could break existing queries.
+    std::string selectColumns;
+    for (const auto& k : columns) {
+      selectColumns += k + ",";
+    }
+
+    // Remove the last comma
+    if (!selectColumns.empty()) {
+      selectColumns.pop_back();
+    }
+
+    hr = pQueryHelper->put_QuerySelectColumns(
+        stringToWstring(selectColumns).c_str());
+    if (FAILED(hr)) {
+      LOG(ERROR) << windowsSearchTableName << ": failed to set columns";
+      return "";
+    }
+  }
+
+  if (!sort.empty()) {
+    hr = pQueryHelper->put_QuerySorting(stringToWstring(sort).c_str());
+    if (FAILED(hr)) {
+      LOG(ERROR) << windowsSearchTableName << ": failed to set sort";
+      return "";
+    }
+  }
+
+  LPWSTR sql;
+  hr = pQueryHelper->GenerateSQLFromUserQuery(
+      stringToWstring(userInput).c_str(), &sql);
+  if (FAILED(hr)) {
+    LOG(ERROR) << windowsSearchTableName
+               << ": failed to generate SQL from user query";
+    return "";
+  }
+
+  std::string ret = wstringToString(sql);
+  CoTaskMemFree(sql);
+  return ret;
+}
+
+QueryData genWindowsSearch(QueryContext& context) {
+  QueryData results;
+  HRESULT hr = CoInitializeEx(0, COINIT_MULTITHREADED);
+  auto const coUninitalizeGuard =
+      scope_guard::create([hr]() { CoUninitialize(); });
+
+  CDataSource cDataSource;
+  hr = cDataSource.OpenFromInitializationString(
+      L"provider=Search.CollatorDSO.1;EXTENDED "
+      L"PROPERTIES=\"Application=Windows\"");
+  if (FAILED(hr)) {
+    LOG(ERROR) << windowsSearchTableName << ": error initializing CDataSource";
+    return results;
+  }
+  auto const cDataSourceGuard =
+      scope_guard::create([&cDataSource]() { cDataSource.Close(); });
+
+  CSession cSession;
+  hr = cSession.Open(cDataSource);
+  if (FAILED(hr)) {
+    LOG(ERROR) << windowsSearchTableName << ": error opening CSession";
+    return results;
+  }
+  auto const cSessionGuard =
+      scope_guard::create([&cSession]() { cSession.Close(); });
+
+  LONG maxResults = 100;
+  if (context.hasConstraint("max_results", EQUALS)) {
+    auto maxResultsConstraint =
+        context.constraints["max_results"].getAll(EQUALS);
+    auto maxResultsStr = SQL_TEXT(*maxResultsConstraint.begin());
+    maxResults = std::stol(maxResultsStr);
+  }
+
+  class windowsToOsqueryColumn {
+   public:
+    std::string osqueryColumnName;
+    std::string osqueryColumnType;
+  };
+
+  std::map<std::string, windowsToOsqueryColumn>
+      defaultWindowsPropertiesToOsqueryColumns = {
+          {"system.itemname", {"name", "TEXT"}},
+          {"system.itempathdisplay", {"path", "TEXT"}},
+          {"system.size", {"size", "INTEGER"}},
+          {"system.datecreated", {"date_created", "INTEGER"}},
+          {"system.datemodified", {"date_modified", "INTEGER"}},
+          {"system.fileowner", {"owner", "TEXT"}},
```

Comment:
Can you explore why [system.fileowner](https://learn.microsoft.com/en-us/windows/win32/properties/props-system-fileowner) is being set to NULL on your latest example output?

Reply:
hmm ... my best guess here is that my system index is not complete. I have rebuilt it once or twice just fiddling with things, it's rebuilding now. However, you can prove that the table is working correctly and the windows api is in fact returning null by running:

```sql
select * from windows_search where query = 'owner:*james*' and owner not like '%james%';
```

I think with this ☝️ query, if the api found something owned by james, but the table read it as null, we would get the result.

Reply:
I reproduced the same issue using [WSOleDB](https://github.com/microsoft/Windows-classic-samples/blob/7cbd99ac1d2b4a0beffbaba29ea63d024ceff700/Samples/Win7Samples/winui/WindowsSearch/WSOleDB/WSOleDB.cpp) sample app.  It seems that there are some files where [system.fileowner](https://learn.microsoft.com/en-us/windows/win32/properties/props-system-fileowner) is not returning the expected indexed data. I think we can leave this as is as it currently working for query time as you pointed out.